python/cuml/metrics/pairwise_distances.pyx

```diff
@@ -0,0 +1,222 @@
+        return DistanceType.EucUnexpandedL1
+    elif metric_str == 'cosine':
+        return DistanceType.EucExpandedCosine
+    elif metric_str == 'euclidean':
+        return DistanceType.EucUnexpandedL2Sqrt
+    elif metric_str == 'haversine':
+        raise ValueError(" The metric: '{}', is not supported at this time."
+                         .format(metric_str))
+    elif metric_str == 'l2':
+        return DistanceType.EucUnexpandedL2Sqrt
+    elif metric_str == 'l1':
+        return DistanceType.EucUnexpandedL1
+    elif metric_str == 'manhattan':
+        return DistanceType.EucUnexpandedL1
+    elif metric_str == 'nan_euclidean':
+        raise ValueError(" The metric: '{}', is not supported at this time."
+                         .format(metric_str))
+    else:
+        raise ValueError("Unknown metric: {}".format(metric_str))
+
+
+@with_cupy_rmm
+def pairwise_distances(X, Y=None, metric="euclidean", handle=None,
+                       convert_dtype=True, **kwds):
+    """
+    Compute the distance matrix from a vector array X and optional Y.
+
+    This method takes either one or two vector arrays, and returns
+    a distance matrix.
+
+    If Y is given (default is None), then the returned matrix is the pairwise
+    distance between the arrays from both X and Y.
+
+    Valid values for metric are:
+
+    - From scikit-learn: ['cityblock', 'cosine', 'euclidean', 'haversine',
+                          'l1', 'l2', 'manhattan'].
+                          Sparse matrices are not supported.
+
+    Parameters
+    ----------
+    X : array-like (device or host) shape = (n_samples_x, n_features)
+        Acceptable formats: cuDF DataFrame, NumPy ndarray, Numba device
+        ndarray, cuda array interface compliant array like CuPy
+
+    Y : array-like (device or host), optional shape = (n_samples_y, n_features)
+        Acceptable formats: cuDF DataFrame, NumPy ndarray, Numba device
+        ndarray, cuda array interface compliant array like CuPy
+
+    metric : string
+        The metric to use when calculating distance between instances in a
+        feature array.
+
+    convert_dtype : bool, optional (default = True)
+        When set to True, the method will, when necessary, convert
+        Y to be the same data type as X if they differ. This
+        will increase memory used for the method.
+
+    Returns
+    -------
+    D : array [n_samples_x, n_samples_x] or [n_samples_x, n_samples_y]
+        A distance matrix D such that D_{i, j} is the distance between the
+        ith and jth vectors of the given matrix X, if Y is None.
+        If Y is not None, then D_{i, j} is the distance between the ith array
+        from X and the jth array from Y.
+    """
+
+    handle = cuml.common.handle.Handle() if handle is None else handle
+    cdef cumlHandle *handle_ = <cumlHandle*> <size_t> handle.getHandle()
+
+    # Determine the input type to convert to when returning
+    output_type = _input_to_type(X)
```

Comment:
It's nice when this is passed as an argument to the function. It gives users better control without having to explicitly convert their inputs.

Reply:
Are you referring to the handle or the input type?

Reply:
we use an `output_type` argument in the constructor of the estimators so the user can override the output type. The default for this value is "input", which mirrors the input value.

Reply:
Here's an example: https://github.com/rapidsai/cuml/blob/branch-0.15/python/cuml/manifold/umap.pyx#L300

Reply:
Ah, I understand now. I looked around at other methods in metrics (and elsewhere) to follow as an example but didn't see any methods that used `ouput_type` as an argument. The only time I have seen `output_type` passed as an argument is if it inherits from `cuml.common.base.Base`. Since `pairwise_distances` is a static method, I defaulted to using "input" as the output type.

I can easily add the `output_type` argument but it will be different from the other methods in `cuml.metrics`. Do we want to keep the method signatures consistent across the module?

Reply:
We should keep the APIs consistent. We recently (a couple releases ago) went through and updated the estimators to use this. I think there might be a couple that still the update but most should be using it. It’s an easy enough addition to the API and it makes a world of difference for users.

Reply:
What do you think about adding it into the pairwise distances and we can open a github issue to eventually scrape through the other convenience functions and update them?

Reply:
I like this idea but could use some input on the implementation. I added a helper function `_determine_stateless_output_type` to python/cuml/common/base.pyx which performs the same steps as the `Base` class to determine the output type for stateless functions (see this commit https://github.com/mdemoret-nv/cuml/commit/b60e76ee57793e245b779a8d59a793e12e2230d7). Couple questions:

1. I added this function because I couldn't find a similar utility. Is there anything else in the code base that performs a similar function? Don't want to duplicate any code.
2. Is this the best place for this function? Doesn't really have much to do with the`Base` class but it seems like an ok place given the other utility functions
3. How do we handle `output_type="input"`? Is it a valid output type? I added a test case to exercise the `output_type` but it fails because the existing utility function `_check_output_type_str` doesn't like "input" as a type string. I tried looking around for other test cases that use this output type but couldn't find much. Not sure how risky it would be to update `_check_output_type_str`. Thoughts?

Reply:
1. I don't believe such a function exists, though I think we should try to have the base use your function or extract the underlying logic into separate functions that both can use. I'm thinking about the case where the logic gets updated in base but not in the stateless version. 

2. Since base is specifically a stateful class, `input_utils` would be a better home for this function, I think.

3. I don't think the function `_check_output_type_str()` should ever be invoked when output_type == input. In fact, the more I looked through the code, I realized "input" is just the global option that sets the default when the user doesn't provide anything. Otherwise, the user-provided output type is used.

Reply:
1. I agree that the Base should use the function as well. The only problem is determining the output type in Base is done in 2 steps (Once in `__init__` and then again once `_set_output_type` is called) and this function just combines those into a single step. So it wouldnt be possible for them to actually share this method. They could share a little bit more code but I think a new issue would be better to look into how `output_type` is handled across the board.
2. If I move it to `input_utils`, then some other methods would need to be moved as well (`_input_to_type`, `_check_output_type_str` and at that point you might as well move `_input_target_to_dtype` too). Is it worth moving them all?
3. Thinking about it more, I agree with you. "input" really only gets set if `output_type == None` and is more of an internal option. So I think its logic is correct now.

Reply:
1. Issue for this is a great idea.

2. I believe those methods are pretty localized to `Base.pyx` so I don't believe moving them should be too difficult. I chose `input_utils` because it already exists but I think those functions might be better off in their own `output_utils.py`

Reply:
Just to clarify, do you mean an issue would be great for just #1 or for #1 and #2?